### PR TITLE
Ports SS220 AutoComplete for Documents

### PR DESCRIPTION
**Pull request by Henri215**
## What Does This PR Do

Port from https://github.com/ss220-space/Paradise/pull/934
image

You can now click a small button located next to the text editing to automaticaly fill it with the following information: Current name, job, signature, account number, PIN, station name, date, time, gender, species.

## Why It's Good For The Game

This will hopefully make filling paperwork more easier for those that are not accustomed to it.
## Images of changes

https://user-images.githubusercontent.com/77684085/220946072-9eed4bf8-bbdf-4d63-a6a8-28b929524eea.mp4

## Testing

- See video above
## Changelog
:cl: Henri215, PhantornRU
add: You can now use the new autocomplete feature when writing on a paper (the small "A" located next to "write")
/:cl:

### CODE REVIEW

File: code/modules/paperwork/paper.dm [hunk 1]
```diff
@@ -251,10 +251,11 @@
 /obj/item/paper/proc/updateinfolinks()
 	info_links = info
 	var/i = 0
-	for(i=1,i<=fields,i++)
-		addtofield(i, "<font face=\"[deffont]\"><A href='?src=[UID()];write=[i]'>write</A></font>", 1)
-	info_links = info_links + "<font face=\"[deffont]\"><A href='?src=[UID()];write=end'>write</A></font>"
-
+	for(i = 1, i <= fields, i++)
```
Comment (Sirryan2002) [inline code/modules/paperwork/paper.dm:254]:
```suggestion
	for(var/i in 1 to fields)
```

File: code/modules/paperwork/paper.dm [hunk 2]
```diff
@@ -251,10 +251,11 @@
...
+		var/write_1 = "<font face=\"[deffont]\"><A href='?src=[UID()];write=[i]'>write</A></font>"
+		var/write_2 = "<font face=\"[deffont]\"><A href='?src=[UID()];auto_write=[i]'><span style=\"color: #409F47; font-size: 10px\">\[A\]</span></A></font>"
+		addtofield(i, "[write_1][write_2]", 1)
+	info_links = info_links + "<font face=\"[deffont]\"><A href='?src=[UID()];write=end'>write</A></font>" + "<font face=\"[deffont]\"><A href='?src=[UID()];auto_write=end'><span style=\"color: #409F47; font-size: 10px\">\[A\]</span></A></font>"
```
Comment (Sirryan2002) [inline code/modules/paperwork/paper.dm:258]:
```suggestion
		var/write_1 = "<font face=\"[deffont]\"><a href='?src=[UID()];write=[i]'>write</a></font>"
		var/write_2 = "<font face=\"[deffont]\"><a href='?src=[UID()];auto_write=[i]'><span style=\"color: #409F47; font-size: 10px\">\[a\]</span></a></font>"
		addtofield(i, "[write_1][write_2]", 1)
	info_links = info_links + "<font face=\"[deffont]\"><a href='?src=[UID()];write=end'>write</a></font>" + "<font face=\"[deffont]\"><a href='?src=[UID()];auto_write=end'><span style=\"color: #409F47; font-size: 10px\">\[a\]</span></a></font>"
```
HTML elements needs to be lowercase

File: code/modules/paperwork/paper.dm [hunk 3]
```diff
@@ -304,53 +305,81 @@
...
+*/
+	input_element = parsepencode(input_element, item_write, usr) // Encode everything from pencode to html
+	if(id != "end")
+		addtofield(text2num(id), input_element) // He wants to edit a field, let him.
+	else
+		info += input_element // Oh, he wants to edit to the end of the file, let him.
+	populatefields()
+	updateinfolinks()
+	item_write.on_write(src, usr)
+	show_content(usr, forceshow = 1, infolinks = 1)
```
Comment (Sirryan2002) [inline code/modules/paperwork/paper.dm:333]:
```suggestion
	show_content(usr, forceshow = TRUE, infolinks = TRUE)
```

File: code/modules/paperwork/paper.dm [hunk 4]
```diff
@@ -304,53 +305,81 @@
...
+/*
+	t = checkhtml(t)
+	// check for exploits
+	for(var/bad in paper_blacklist)
+		if(findtext(t,bad))
+			to_chat(usr, "<span class='notice'>You think to yourself, \</span>"Hm.. this is only paper...\"")
+			log_admin("PAPER: [key_name(usr)] tried to use forbidden word in [src]: [bad].")
+			message_admins("PAPER: [key_name_admin(usr)] tried to use forbidden word in [src]: [bad].")
+			return
+*/
```
Comment (Sirryan2002) [inline code/modules/paperwork/paper.dm:324]:
```suggestion
```

File: code/modules/paperwork/paper.dm [hunk 5]
```diff
@@ -304,53 +305,81 @@
...
-		updateinfolinks()
-
-		i.on_write(src,usr)
-
-		show_content(usr, forceshow = 1, infolinks = 1)
-
-		update_icon()
-
+		// var/t = strip_html_simple(input(usr, "What text do you wish to add to " + (id=="end" ? "the end of the paper" : "field "+id) + "?", "[name]", null),8192) as message
+		// var/t =  strip_html_simple(input("Enter what you want to write:", "Write", null, null)  as message, MAX_MESSAGE_LEN)
```
Comment (Sirryan2002) [inline code/modules/paperwork/paper.dm:380]:
```suggestion
```

File: code/modules/paperwork/paper.dm [hunk 6]
```diff
@@ -304,53 +305,81 @@
 		\[time\] : Inserts the current station time in HH:MM:SS.<br>
 	</BODY></HTML>"}, "window=paper_help")
 
+/obj/item/paper/proc/topic_href_write(id, input_element)
+	var/obj/item/item_write = usr.get_active_hand() // Check to see if he still got that darn pen, also check if he's using a crayon or pen.
+	add_hiddenprint(usr) // No more forging nasty documents as someone else, you jerks
+	if(!istype(item_write, /obj/item/pen) && !istype(item_write, /obj/item/toy/crayon))
+		return
+	if(src.loc != usr && !src.Adjacent(usr) && !((istype(src.loc, /obj/item/clipboard) || istype(src.loc, /obj/item/folder)) && (src.loc.loc == usr || src.loc.Adjacent(usr))))
```
Comment (Sirryan2002) [inline code/modules/paperwork/paper.dm:313]:
```suggestion
	if(loc != usr && !Adjacent(usr) && !((istype(loc, /obj/item/clipboard) || istype(loc, /obj/item/folder)) && (loc.loc == usr || loc.Adjacent(usr))))
```

File: code/modules/paperwork/paper.dm (hunk 4 above)
Comment (Henri215) [reply code/modules/paperwork/paper.dm:324]: Didnt know if i should keep it or not, removing it.

**Comment by AffectedArc07** [comment]
```
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: Judgement (/obj/item/paper/Court)
  src.loc: the floor (207,124,1) (/turf/simulated/floor/plasteel/dark)
  call stack:
  Judgement (/obj/item/paper/Court): updateinfolinks
  Judgement (/obj/item/paper/Court): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: Monitor Decryption Key (/obj/item/paper/monitorkey)
  src.loc: the floor (210,[135](https://github.com/ParadiseSS13/Paradise/actions/runs/4255532702/jobs/7403267307#step:9:136),1) (/turf/simulated/floor/plasteel)
  call stack:
  Monitor Decryption Key (/obj/item/paper/monitorkey): updateinfolinks
  Monitor Decryption Key (/obj/item/paper/monitorkey): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: the paper- \'Standard Operatin... (/obj/item/paper/sop)
  src.loc: the carpet (234,[145](https://github.com/ParadiseSS13/Paradise/actions/runs/4255532702/jobs/7403267307#step:9:146),1) (/turf/simulated/floor/carpet)
  call stack:
  the paper- \'Standard Operatin... (/obj/item/paper/sop): updateinfolinks
  the paper- \'Standard Operatin... (/obj/item/paper/sop): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: Dissection Guide (/obj/item/paper/abductor)
  src.loc: the alien floor (121,[160](https://github.com/ParadiseSS13/Paradise/actions/runs/4255532702/jobs/7403267307#step:9:161),1) (/turf/simulated/floor/plating/abductor)
  call stack:
  Dissection Guide (/obj/item/paper/abductor): updateinfolinks
  Dissection Guide (/obj/item/paper/abductor): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: Dissection Guide (/obj/item/paper/abductor)
  src.loc: the alien floor (135,160,1) (/turf/simulated/floor/plating/abductor)
  call stack:
  Dissection Guide (/obj/item/paper/abductor): updateinfolinks
  Dissection Guide (/obj/item/paper/abductor): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: Teleporter Instructions (/obj/item/paper/synditele)
  src.loc: the shuttle floor (186,[162](https://github.com/ParadiseSS13/Paradise/actions/runs/4255532702/jobs/7403267307#step:9:163),1) (/turf/simulated/floor/mineral/plastitanium/red)
  call stack:
  Teleporter Instructions (/obj/item/paper/synditele): updateinfolinks
  Teleporter Instructions (/obj/item/paper/synditele): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: Dissection Guide (/obj/item/paper/abductor)
  src.loc: the alien floor (121,[174](https://github.com/ParadiseSS13/Paradise/actions/runs/4255532702/jobs/7403267307#step:9:175),1) (/turf/simulated/floor/plating/abductor)
  call stack:
  Dissection Guide (/obj/item/paper/abductor): updateinfolinks
  Dissection Guide (/obj/item/paper/abductor): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: Dissection Guide (/obj/item/paper/abductor)
  src.loc: the alien floor (135,174,1) (/turf/simulated/floor/plating/abductor)
  call stack:
  Dissection Guide (/obj/item/paper/abductor): updateinfolinks
  Dissection Guide (/obj/item/paper/abductor): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: the paper- \'Memo\' (/obj/item/paper/syndimemo)
  src.loc: the floor (192,[177](https://github.com/ParadiseSS13/Paradise/actions/runs/4255532702/jobs/7403267307#step:9:178),1) (/turf/simulated/floor/plasteel)
  call stack:
  the paper- \'Memo\' (/obj/item/paper/syndimemo): updateinfolinks
  the paper- \'Memo\' (/obj/item/paper/syndimemo): New
[2023-02-23T18:10:08] Runtime in paper.dm,253: bad number
  proc name: updateinfolinks (/obj/item/paper/proc/updateinfolinks)
  src: the our demands (/obj/item/paper/crumpled/bloody/ruins/lavaland/clown_planet)
  src.loc: the carpet ([192](https://github.com/ParadiseSS13/Paradise/actions/runs/4255532702/jobs/7403267307#step:9:193),247,1) (/turf/simulated/floor/carpet/purple)
  call stack:
  the our demands (/obj/item/paper/crumpled/bloody/ruins/lavaland/clown_planet): updateinfolinks
  the our demands (/obj/item/paper/crumpled/bloody/ruins/lavaland/clown_planet): New
  (This error will now be silenced for 10 minutes)
```

File: code/modules/paperwork/paper.dm [hunk 7]
```diff
@@ -304,53 +304,69 @@
 		\[time\] : Inserts the current station time in HH:MM:SS.<br>
 	</BODY></HTML>"}, "window=paper_help")
 
+/obj/item/paper/proc/topic_href_write(id, input_element)
+	var/obj/item/item_write = usr.get_active_hand() // Check to see if he still got that darn pen, also check if he's using a crayon or pen.
+	add_hiddenprint(usr) // No more forging nasty documents as someone else, you jerks
+	if(!istype(item_write, /obj/item/pen) && !istype(item_write, /obj/item/toy/crayon))
+		return
+	if(loc != usr && !Adjacent(usr) && !((istype(loc, /obj/item/clipboard) || istype(loc, /obj/item/folder)) && (loc.loc == usr || loc.Adjacent(usr))))
```
Comment (S34NW) [inline code/modules/paperwork/paper.dm:312]: Instead of `loc.loc`, perhaps you could check if usr is in the turf of the paper?

**Review by S34NW — APPROVED** [review]

**Review by AffectedArc07 — APPROVED** [review]

**Pull request merged**